---
title: "Breastcancer Gene Expression Study: KPNA2 gene"
author: "Lieven Clement"
output: html_document
---

#Background
Background: Histologic grade in breast cancer provides clinically important prognostic information. Researchers examined whether histologic grade was associated with gene expression profiles of breast cancers and whether such profiles could be used to improve histologic grading. In this tutorial we will assess the impact of histologic grade on expression of the KPNA2 gene that is known to be associated with poor BC prognosis.
The patients, however, do not only differ in the histologic grade, but also on their lymph node status. The lymph nodes were not affected (0) or chirugically removed (1).


#Data analysis
##Import KPNA2 data in R
```{r}
kpna2=read.table("gse2990BreastcancerOneGene.txt",header=TRUE)
head(kpna2)
```

##Transform the variable grade and node to a factor
```{r}
kpna2$grade=as.factor(kpna2$grade) 
kpna2$node=as.factor(kpna2$node)
```

##Transform the variable grade and node to a factor
```{r}
kpna2$logExp=log2(kpna2$gene)
```

##Data exploration
Histological grade and lymph node status can both have an effect on the KPNA2 gene. Moreover, it is also possible that the effect of the histological grade changes according to the lymph node status. Therefore, we plot the gene expression for each grade x node combination. 

```{r}
#Note, that we recast node.f:grade into a double to avoid that boxplots are plotted
#We suppress the x axis to be able to plot the labels ourselves.
kpna2$gradeNode=factor(paste0("n",kpna2$node,"g",kpna2$grade))
plot(logExp~gradeNode,xlab="node x grade",ylab="Intensity (log2)",data=kpna2)
```

The plot suggests

- an effect of the histological grade
- an effect of node status
- The differential expression associated to grade seems to differ according to the lymph node status (interaction) 


##Hypotheses of interest
- KPNA2 expression is on average different between grade 3 and grade 1 in patients with unaffected lymph nodes
- KPNA2 expression is on average different between grade 3 and grade 1 in patients where lymph nodes were affected and chirugically removed
- KPNA2 expression is on average different between patients with affected and unaffected lymph nodes and a tumor of grade 1
- KPNA2 expression is on average different between patients with affected and unaffected lymph nodes and a tumor of grade 3
- The fold change of the KPNA2 gene between grade 3 and grade 1 differs according to the lymph node status. 

##Model
Histologic grade and lymph node status can have an effect on the kpna2 gene. Moreover, it is also possible that the differential expression due to histological grade is differt in patients that have unaffected lymph nodes and patients for which the lymph nodes had to be removed. Hence, we will have to model the gene expression by using main effects for grade, node and a grade x node interaction. 
 
```{r}
#Model with main effects for histological grade and node and grade x node interaction
fit=lm(logExp~grade+node+grade:node,data=kpna2)
plot(fit)
```

The variance is more or less equal for every treatment x node combination. 
The QQ-plot of the residuals shows no severe deviations from normality.

```{r}
summary(fit)
#Calculate confidence intervals for parameters of model
CIfit=confint(fit)
#log2 FC between g3n0-g1n0, g1n1-g1n0
#and log2 difference in FC g3n1-g1n1 and FC g3n0-g1n0
CIfit
#Transform parameters and the CI back to the original scale
2^fit$coef
2^CIfit
```

#Interpretation of model parameters and statistical tests
We model the log$_2$-transformed intensities with the following model: 
$$
y=\beta_0+\beta_{g3}x_{g3}+\beta_{n1}x_{n1}+\beta_{g3n1}x_{g3}x_{n1},
$$

with $\beta_0$ the intercept, $\beta_{g3}$ the main effect for grade, $x_{g3}$ a dummy variable for grade which is 0 for the control treatment in the absence of grade and 1 for the treatment with grade, $\beta_{n1}$ the main effect for node, $x_{n1}$ a dummy variable that is 0 for the measurements of patients with unaffected lymph nodes and 1 for patients for which the lymph nodes were removed and $\beta_{g3n1}$ the interaction effect between grade and node. 
To ease the interpertation of the parameters, $\log_2$ transformed mean intensities are given for each treatment group as well as corresponding contrasts between treatments, which have an interpretation in terms of $\log_2$ transformed fold changes (FC). 

- $\log_2\hat \mu_{g1n0}=\hat\beta_0$, $\log_2 \hat\mu_{g3n0}=\hat\beta_0+\hat\beta_{g3}$ --> $\log_2 \widehat{FC}_{g3n0-g1n0}=\hat\beta_{g3}$

- $\log_2 \hat \mu_{g1n1}=\hat\beta_0+\hat\beta_t$, $\log_2 \hat \mu_{g3n1}=\hat\beta_0+\hat\beta_{g3}+\hat\beta_t+\hat\beta_{g3n1}$ --> $\log_2 \widehat{FC}_{g3n1-g1n1}=\hat \beta_{g3} +\hat\beta_{g3n1}$

- $\log_2 \widehat{FC}_{g1n1-g1n0}=\hat\beta_{n1}$, $\log_2 \widehat{FC}_{g3n1-g3n0}=\hat\beta_{n1}+\hat\beta_{g3n1}$ --> $\log_2\frac{\widehat{FC}_{g3n1-g1n1}}{\widehat{FC}_{g3n0-g1n0}}=\hat\beta_{g3n1}$

with $\log_2\hat \mu_{g1n0}$, $\log_2\hat \mu_{g3n0}$, $\log_2\hat \mu_{g1n1}$ and $\log_2\hat \mu_{g3n1}$ the estimated mean $\log_2$ transformed intensity for patients with grade 1 and node 0 status, grade 3 and node 0 status, grade 1 and node 1 status and grade 3 and node 1 status, respectively. With $\log_2 \widehat{FC}_{b-a}$ we indicate $\log_2$ transformed fold change estimates between treatment b and treatment a, i.e. $\log_2 \widehat{FC}_{b-a}=\log_2 \hat\mu_{b}-\log_2 \hat\mu_a=\log_2 \frac{\hat\mu_{b}}{\hat\mu_{a}}$.

The model immediately provides statistical tests for assessing the significance of fold changes between grade 3 and grade 1 for patients with unaffected lymph nodes (n=0) $\log_2 \widehat{FC}_{g3n0-g1n0}$,  fold changes between the grade 1-node 1 patients and grade 1- node 0 patients $\log_2 \widehat{FC}_{g1n1-g3n0}$ and for differences in fold change due to histological grade for node 1 patients and node 0 patients. $\log_2\frac{\widehat{FC}_{g3n1-g1n1}}{\widehat{FC}_{g3n0-g1n0}}$, the interaction term. 
```{r}
summary(fit)
```

- The average intensity for grade 1 patients with unaffected lymph nodes equals $\exp(\hat \beta_0)$=
`r round(2^fit$coef["(Intercept)"],2)`.
- The hystological grade effect in patients with unaffected lymph nodes and lymph node effect in grade 1 patients are extremely significant (p<<0.001) and very significant (p=0.002), respectively. 
	- When lymph nodes are unaffected, the expression is on average `r round(2^fit$coef["grade3"],2)` times higher for patients with histological grade 3 than patients with histological grade 1 (95% CI [`r round(2^CIfit["grade3",],2)`]). 
	- The gene expression in histological grade 1 patients is on average `r round(1/2^fit$coef["node1"],2)` times higher when the lymph nodes are affected (95% CI [`r round(1/2^CIfit["node1",],2)`]).
- The difference in histological grade effect between patients with and without affected lymph nodes is very significant (p=0.007). The fold change due to histological grade is on average `r round(2^fit$coef["grade3:node1"],2)` times lower in patients with affected lymph nodes (95% CI [`r round(2^CIfit["grade3:node1",],2)`]).


##Histological grade effect in patients with affected lymph nodes. 
The researchers, however, are also interested in testing the significance of the fold change between patients with grade 3 and grade 1 tumors that had their lymph nodes removed This involves assessing a contrast of the model parameters: $\log_2 \widehat{FC}_{g3n1-g1n1}=\hat \beta_{g3} +\hat\beta_{g3n1}$.

###Setup Contrast
 We will first setup the contrast. 
```{r}
#number of parameters in the model
p=length(fit$coef)
L=matrix(0,nrow=1,ncol=p)
#give the rows and columns of the matrix interpretable names
colnames(L)=names(fit$coef)
rownames(L)="FCg3n1_g1n1"
L[1,2]=1
L[1,4]=1
L
```

### Test for contrast
We can assess the contrast ourself by extracting the variance-covariance matrix. 
Alternatively, the multcomp package can be used.

```{r}
library(multcomp)
########################
#perform test for contrast at node 48 h
########################
testFCg3n1_g1n1=glht(fit, linfct = L)
summary(testFCg3n1_g1n1)
##########################
```

### Confidence intervals FC for effect of grade in patients with removed lymph nodes.
```{r}
#Calculate confidence intervals for FC for grade effect with removed lymph nodes
CIFCg3n1_g1n1=confint(testFCg3n1_g1n1)
# log2 FC on T48
CIFCg3n1_g1n1
#FC  for FC on T48
2^CIFCg3n1_g1n1$confint
```

The histological grade effect in patients with affected lymph nodes is also extremely significant (p<<0.001).
The expression of the kpna2 gene in patients with affected lymph nodes is on average `r  round(2^CIFCg3n1_g1n1$confint,2)[1]` times higher when the histological grade 3 (95% CI [`r round(2^CIFCg3n1_g1n1$confint,2)[-1]`]). 


##Effect of lymph node status in patients with grade 3 tumor. 
The researchers, however, are also interested in assessing the difference in expression between patient with removed lymph nodes and unaffected lymph nodes when they both have a grade 3 tumor. This involves  a contrast of the model parameters: $\log_2 \widehat{FC}_{g3n1-g3n0}=\hat \beta_{n1} +\hat\beta_{g3n1}$.

###Setup Contrast
 We will first setup the contrast. 
```{r}
#number of parameters in the model
p=length(fit$coef)
L=matrix(0,nrow=1,ncol=p)
#give the rows and columns of the matrix interpretable names
colnames(L)=names(fit$coef)
rownames(L)="FCg3n1_g1n1"
L[1,3]=1
L[1,4]=1
L
```

### Test for contrast
We can assess the contrast ourself by extracting the variance-covariance matrix. 
Alternatively, the multcomp package can be used.

```{r}
library(multcomp)
########################
#perform test for contrast at node 48 h
########################
testFCg3n1_g3n0=glht(fit, linfct = L)
summary(testFCg3n1_g3n0)
##########################
```

### Confidence intervals FC for effect of grade in patients with removed lymph nodes.
```{r}
#Calculate confidence intervals for FC for grade effect with removed lymph nodes
CIFCg3n1_g3n0=confint(testFCg3n1_g3n0)
# log2 FC on T48
CIFCg3n1_g3n0
#FC  for FC on T48
2^CIFCg3n1_g3n0$confint
```

The gene expression for patients with removed lymph nodes and a grade 3 tumor are on average not significantly different from patients with unaffected lymph nodes and a grade 3 tumor (p=`r round(summary(testFCg3n1_g3n0)$test$pvalues[1],3)`). 

###Assess hypotheses of interest and correct for multiple testing
This can be done automatically in the package multcomp. We will also assess the impact of the lymph node status
```{r}
#######################################################################
#Assess all 3 hypotheses of interest and correct for multiple testing
#######################################################################
#Construct contrast matrix with 3 contrasts
L=matrix(0,nrow=5,ncol=p)
#give columns and rows of L sensible names
colnames(L)=names(fit$coef)
rownames(L)=c("FCg3n0_g1n0","FCg3n1_g1n1","FCg1n1_g1n0","FCg3n1_g3n0","Interaction")
#contrast for FC at node 10
L[1,2]=1
#contrast for FC at node 48
L[2,c(2,4)]=1
#contrast for lymph node status in patients 
L[3,3]=1
L[4,3]=L[4,4]=1
#contrast for interaction
L[5,4]=1
L

#Perform test
testFCAll=glht(fit, linfct = L)
summary(testFCAll)
#Calculate confidence intervals on log2 FC and log2 difference in FC between T48 and T10
CIFCAll=confint(testFCAll)
CIFCAll
#Calculate FC and FC factor between T48 and T10
2^CIFCAll$confint
```

#Matrix implementation
##Parameter estimation
Parameter estimator equals
$$\hat\beta=(X^T X)^{-1}X^TY$$
```{r}
X=model.matrix(~grade*node,data=kpna2)
betaMe=solve(t(X)%*%X)%*%t(X)%*%kpna2$logExp
betaMe
fit$coefficients
range(betaMe-fit$coefficients)
```
##Standard error
The variance on the model parameters equals
$$\Sigma_\hat\beta=(X^TX)^{-1}\hat\sigma^2$$
```{r}
fitMe=X%*%betaMe
res=kpna2$logExp-fitMe
n=nrow(X)
p=ncol(X)
varY=sum(res^2)/(n-p)
varBeta=solve(t(X)%*%X)*varY
se=sqrt(diag(varBeta))
tStat=betaMe/se
pVal=pt(-abs(tStat),df=n-p)*2
summary(fit)$coef
cbind(estimate=betaMe,se=se,tStat=tStat,pVal=pVal)
```
##Contrast
```{r}
contrastMe <- L%*%betaMe
contrastMeVar <- L%*%varBeta%*%t(L)
contrastMeSe <- sqrt(diag(contrastMeVar))
contrastMeTstat <- contrastMe/contrastMeSe
contrastMeP <- pt(-abs(contrastMeTstat),df=n-p)*2
summary(testFCAll,test=adjusted("none"))
cbind(estimate=contrastMe,se=contrastMeSe,tStat=contrastMeTstat,pval=contrastMeP)
```
#Conclusions
There is an extremely significant association between the histological grade and the expression of the KPNA2 gene in patients with unaffected lymph nodes as well as in patients with removed lymph nodes (both $p<0.001$). 
On average, the fold change between histological grade 3 and grade 1 patients varied form `r round(2^CIFCAll$confint["FCg3n1_g1n1",1],2)` (95% CI [`r round(2^CIFCAll$confint["FCg3n1_g1n1",2:3],2)`]) in patients with removed lymph nodes upto `r round(2^CIFCAll$confint["FCg3n0_g1n0",1],2)` (95% CI [`r round(2^CIFCAll$confint["FCg3n0_g1n0",2:3],2)`]) for patients with unaffected lymph nodes, respectively.  
The effect of histological grade between patients with affected lymph nodes also differs significantly from that of patients with unaffected lymph nodes (p=`r round(summary(testFCAll)$test$pvalues[5],3)`). The fold change due to the histological grade is on average `r round(2^-CIFCAll$confint["Interaction",1],2)` higher in patients with unaffected lymph nodes than in patients with affected lymph nodes (95% CI [`r round(2^-CIFCAll$confint["Interaction",3:2],2)`] ). (All p-values and confidence intervals were adjusted for multiple testing). 

The gene expression in patients with a histological grade 1 tumor on average differs very significantly according to their lymph node status (p=`r round(summary(testFCAll)$test$pvalues[3],3)`). 
For patients with a tumor of histological grade 1, the gene expression of the KPNA2 gene is on average `r round(2^CIFCAll$confint["FCg1n1_g1n0",1],2)` times higher in patients with removed lymph nodes than in patients with unaffected lymph nodes (95% CI[`r round(2^CIFCAll$confint["FCg1n1_g1n0",2:3],2)`]).
The average gene expression in patients with a histological grade 3 tumor does not differ significantly according to their lymph node status (p=`r round(summary(testFCAll)$test$pvalues[4],3)`). 


